File: code/java/ch12/src/main/java/com/github/zxh0/luago/state/LuaStack.java
package com.github.zxh0.luago.state;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import static com.github.zxh0.luago.api.LuaState.LUA_REGISTRYINDEX;

class LuaStack {

    /* virtual stack */
    private final ArrayList<Object> slots = new ArrayList<>();
    /* call info */
    LuaStateImpl state;
    Closure closure;
    List<Object> varargs;
    Map<Integer, UpvalueHolder> openuvs;
    int pc;
    /* linked list */
    LuaStack prev;

    int top() {
        return slots.size();
    }

    void push(Object val) {
        if (slots.size() > 10000) { // TODO
            throw new StackOverflowError();
        }
        slots.add(val);
    }

    Object pop() {
        return slots.remove(slots.size() - 1);
    }

    void pushN(List<Object> vals, int n) {
        int nVals = vals == null ? 0 : vals.size();
        if (n < 0) {
            n = nVals;
        }
        for (int i = 0; i < n; i++) {
            push(i < nVals ? vals.get(i) : null);
        }
    }

    List<Object> popN(int n) {
        List<Object> vals = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            vals.add(pop());
        }
        Collections.reverse(vals);
        return vals;
    }

    int absIndex(int idx) {
        return idx >= 0 || idx <= LUA_REGISTRYINDEX
                ? idx : idx + slots.size() + 1;
    }

    boolean isValid(int idx) {
        if (idx < LUA_REGISTRYINDEX) { /* upvalues */
            int uvIdx = LUA_REGISTRYINDEX - idx - 1;
            return closure != null && uvIdx < closure.upvals.length;
        }
        if (idx == LUA_REGISTRYINDEX) {
            return true;
        }
        int absIdx = absIndex(idx);
        return absIdx > 0 && absIdx <= slots.size();
    }

    Object get(int idx) {
        if (idx < LUA_REGISTRYINDEX) { /* upvalues */
            int uvIdx = LUA_REGISTRYINDEX - idx - 1;
            if (closure != null
                    && closure.upvals.length > uvIdx
                    && closure.upvals[uvIdx] != null) {
                return closure.upvals[uvIdx].get();
            } else {
                return null;
            }
        }
        if (idx == LUA_REGISTRYINDEX) {
            return state.registry;
        }
        int absIdx = absIndex(idx);
        if (absIdx > 0 && absIdx <= slots.size()) {
            return slots.get(absIdx - 1);
        } else {
            return null;
        }
    }

    void set(int idx, Object val) {
        if (idx < LUA_REGISTRYINDEX) { /* upvalues */
            int uvIdx = LUA_REGISTRYINDEX - idx - 1;
            if (closure != null
                    && closure.upvals.length > uvIdx
                    && closure.upvals[uvIdx] != null) {
                closure.upvals[uvIdx].set(val);
            }
            return;
        }
        if (idx == LUA_REGISTRYINDEX) {
            state.registry = (LuaTable) val;
            return;
        }
        int absIdx = absIndex(idx);
        slots.set(absIdx - 1, val);
    }

    void reverse(int from, int to) {
        Collections.reverse(slots.subList(from, to + 1));
    }

}

File: code/csharp/ch13/luavm/state/LuaStack.cs
using System;
using System.Collections;
using System.Collections.Generic;
using luavm.api;

namespace luavm.state
{
    public class LuaStack
    {
        internal readonly ArrayList Slots = new ArrayList();
        internal LuaStack Prev;
        internal Closure Closure;
        internal ArrayList Varargs;
        internal int Pc;
        internal LuaState State;
        internal Dictionary<int, Upvalue> Openuvs;

        internal int Top()
        {
            return Slots.Count;
        }

//        internal void check(int n)
//        {
//            var free = slots.Length - top;
//            var slotList = slots.ToList();
//            for (var i = free; i < n; i++)
//            {
//                slotList.Add(null);
//            }
//
//            slots = slotList.ToArray();
//        }

        internal void Push(object val)
        {
            if (Slots.Count > Slots.Capacity)
            {
                // TODO
                throw new StackOverflowException();
            }

            Slots.Add(val);
        }

        internal int AbsIndex(int idx)
        {
            return idx >= 0 || idx <= Constant.LUA_REGISTRYINDEX
                ? idx
                : idx + Slots.Count + 1;
        }

        internal bool IsValid(int idx)
        {
            if (idx < Constant.LUA_REGISTRYINDEX)
            {
                /* upvalues */
                var uvIdx = Constant.LUA_REGISTRYINDEX - idx - 1;
                return Closure != null && uvIdx < Closure.Upvals.Length;
            }

            if (idx == Constant.LUA_REGISTRYINDEX)
            {
                return true;
            }

            var absIdx = AbsIndex(idx);
            return absIdx > 0 && absIdx <= Slots.Count;
        }

        internal object Get(int idx)
        {
            if (idx < Constant.LUA_REGISTRYINDEX)
            {
                /* upvalues */
                var uvIdx = Constant.LUA_REGISTRYINDEX - idx - 1;
                if (Closure != null
                    && Closure.Upvals.Length > uvIdx
                    && Closure.Upvals[uvIdx] != null)
                {
                    return Closure.Upvals[uvIdx].Get();
                }

                return null;
            }

            if (idx == Constant.LUA_REGISTRYINDEX)
            {
                return State.Registry;
            }

            var absIdx = AbsIndex(idx);
            if (absIdx > 0 && absIdx <= Slots.Count)
            {
                return Slots[absIdx - 1];
            }

            return null;
        }

        internal object Pop()
        {
            var v = Slots[Slots.Count - 1];
            Slots.RemoveAt(Slots.Count - 1);
            return v;
        }

        internal void PushN(ArrayList vals, int n)
        {
            var nVals = vals?.Count ?? 0;
            if (n < 0)
            {
                n = nVals;
            }

            for (var i = 0; i < n; i++)
            {
                Push(i < nVals ? vals[i] : null);
            }
        }

        internal ArrayList PopN(int n)
        {
            var vals = new ArrayList(n);
            for (var i = 0; i < n; i++)
            {
                vals.Add(Pop());
            }

            vals.Reverse();
            return vals;
        }


        internal void Set(int idx, object val)
        {
            if (idx < Constant.LUA_REGISTRYINDEX)
            {
                /* upvalues */
                var uvIdx = Constant.LUA_REGISTRYINDEX - idx - 1;
                if (Closure != null
                    && Closure.Upvals.Length > uvIdx
                    && Closure.Upvals[uvIdx] != null)
                {
                    Closure.Upvals[uvIdx].Set(val);
                }

                return;
            }

            if (idx == Constant.LUA_REGISTRYINDEX)
            {
                State.Registry = (LuaTable) val;
                return;
            }

            var absIdx = AbsIndex(idx);
            Slots.RemoveAt(absIdx - 1);
            Slots.Insert(absIdx - 1, val);
        }

        internal void Reverse(int from, int to)
        {
            Slots.Reverse(from, to - from + 1);
            // Collections.reverse(slots.subList(from, to + 1));
        }
    }
}